SwiftRuntimeLibrary/Exceptions.cs

```diff
@@ -3,7 +3,7 @@
 
 using System;
 namespace SwiftRuntimeLibrary {
-	internal static class Exceptions {
+	public static class Exceptions {
```

Comment:
This helper method cannot be inlined since it throws. It means all null checks (cheap) now requires a call (less cheap).

That might not be obvious to callers, which makes it a _not so good_ candidate to be a public API where the impact cannot be known.

OTOH it could be fixed by moving the actual throws to another method... which would make inlining possible.

Reply:
If I do that then it obviates the benefit for using this.

Reply:
Which benefit ?

The source/call remains identical `Exceptions.ThrowOnNull (foo, "foo");` on the caller side.

It's the `ThrowOnNull` implementation that changes, being split in two. The first part only check for null and call a `Throw*` method if needed. Since the method won't be throwing (by itself) it can be inlined (at the JIT's discretion).

If an exception occurs then a call will be needed - but that's already in a non-performant path so an additional call is unlikely to be noticeable.

Reply:
The benefit is to make it more compact. There are some internal places where I use it specifically for the fluent implementation (eg, `base(Exceptions.ThrowOnNull(arg, nameof (arg)),...)` but if this takes away inlining in the common case, there's no point.

Reply:
```
base(Exceptions.ThrowOnNull(arg, nameof (arg)),...)
```

hmm... that would still work with what I said.

IOW the `ThrowOnNull` API (at least its signature) does not have to change.

But it's implementation does need to change. If it does not throw then it can be inlined, but it does not mean the throwing goes back to the caller. There just need to be another (private) method that does the throw.

```csharp
// this method does not throw and the JIT can (at its discretion) inline it
// that can be important since a `call` can be much costlier than a `null` check
public static T ThrowOnNull<T> (T o, string name, string message = null) where T : class
 {
	if (o == null)
		ThrowNullException (name ?? "::no name supplied::", message);
	return o;
}

// this method does throw and the JIT cannot inline it
// however it's not inside the normal, where performance might matter, path
static void ThrowNullException (string name, string message)
{
	if (message == null)
		throw new ArgumentNullException (name);
	else
		throw new ArgumentNullException (name, message);
}
```

Reply:
OK - I see that. I can always add that to the implementation later.